### PR TITLE
Improve Renovate configuration

### PR DESCRIPTION
**Pull request by HonkingGoose**
### Updates to account for Renovate 36 release: [^1]

- Rename `config:base` to `config:recommended`
- Rename `stabilityDays` to `minimumReleaseAge` [^4]
- Use `':prConcurrentLimitNone'` so there's no limit to the number of open PRs at one time
- Limit PR creation to 2 per hour with `':prHourlyLimit2'`

### Improvements to your workflow:

- Drop `':dependencyDashboard',` because it's included in `config:recommended` [^2]
- Instead of disabling Renovate for certain `major` type updates, use the Dependency Dashboard Approval workflow [^3]
- Use `dependencyDashboardHeader` to remind Mastodon maintainers to check the `renovate.json5` file before approving any updates from the Dependency Dashboard

## Context:

I'm combining a small housekeeping update with a **big change** to your Renovate update workflow. I see you're currently _disabling_ Renovate for certain `major` type updates. I recommend you instead use the Dependency Dashboard workflow. [^3]

Pinging @renchap, because they discussed some problems on the Renovate repository recently. They also updated your Renovate config recently. So I want to make sure they see my PR. 😉 

### Benefits of new workflow:

- You can always see `major` updates in your Dependency Dashboard, and can get them when needed
- You always see if you're using old dependencies in your Dependency Dashboard, right now you don't see some `major` updates at all

### Drawbacks:

- You may accidentally select a `major` update that breaks stuff from the Dependency Dashboard
- The Dependency Dashboard will show more updates (because you're currently not seeing those at all...)

[^1]: [Renovate docs, release notes for 36](https://docs.renovatebot.com/release-notes-for-major-versions/#version-36)
[^2]: [Renovate docs, `config:recommended` preset](https://docs.renovatebot.com/presets-config/#configrecommended)
[^3]: [Renovate docs, require approval for `major` updates](https://docs.renovatebot.com/key-concepts/dashboard/#require-approval-for-major-updates)
[^4]: [Renovate docs, `minimumReleaseAge`](https://docs.renovatebot.com/configuration-options/#minimumreleaseage)

### CODE REVIEW

**Comment by renchap** [comment]
Thanks @HonkingGoose, those improvement makes sense.

Is it possible to add a comment to the dependency dashboard somehow to explain why those versions require manual approval? For example to explain that the Ruby version bumps need to be synced with `.ruby-version`?

Also at the moment we are really blocked by https://github.com/renovatebot/renovate/issues/23527 which prevents most of our JS dep updates to be merged, do you know if this is planned to be fixed soon?

**Comment by HonkingGoose** [comment]
> Is it possible to add a comment to the dependency dashboard somehow to explain why those versions require manual approval? For example to explain that the Ruby version bumps need to be synced with `.ruby-version`?

@renchap Just checking: you want the Dependency Dashboard to show some explanatory text why a update requires manual approval? I think there's no way to do that, right now. Please open a Discussion on the Renovate repository to request this feature, and get the maintainers insight on this. Maybe they have an idea on a hackjob-way to get some text in the PR title, but editing the PR title may cause other problems... 😄 

> Also at the moment we are really blocked by [renovatebot/renovate#23527](https://github.com/renovatebot/renovate/issues/23527) which prevents most of our JS dep updates to be merged, do you know if this is planned to be fixed soon?

I don't know. I recommend subscribing to that issue, so you know about any changes when they happen.

**Comment by HonkingGoose** [comment]
@renchap do you want me to make any changes to this PR? Or are you happy to merge as-is? 🙂

File: .github/renovate.json5
Comment (renchap) [inline .github/renovate.json5:7]: Could you remove this?

As there are some PRs we cannot merge easily (due to the Renovate bug above, or needing more work/testing), we would prefer not have a concurrency limit, but still have a rate-limit

File: .github/renovate.json5
Comment (renchap) [inline .github/renovate.json5:49]: For `rack`, we do not want to update it at all. This is an indirect dependency from `rails` (and maybe other), which depends on a specific major version of Rack.

I am wondering if we could remove it entirely, if Renovate is smart enough to detect that we can not bump it due to the major version constraint in `rails`?

File: .github/renovate.json5
Comment (renchap) [inline .github/renovate.json5:54]: Same here, `sidekiq` depends on `redis < 5`. I wonder if Renovate will offer the update as we have a dependency preventing upgrading to it.

**Review by renchap — COMMENTED** [review]
I am generally fine with the changes and added a few comments.

Requiring approval for majors is good, but I fear that someone may trigger a PR for those that really need some thinking without understanding why approval was required (which in indicated in a comment, but not visible from the dashboard).

File: .github/renovate.json5
Comment (HonkingGoose) [reply .github/renovate.json5]: Instead of removing the line, I've replaced it with the `:prConcurrentLimitNone` preset. I've added the `:prHourlyLimit2` (which is the default setting!) to make the config easier to understand.

File: .github/renovate.json5
Comment (HonkingGoose) [reply .github/renovate.json5:50]: I created a preset that groups common Ruby on Rails dependencies into one PR, see:

- https://github.com/renovatebot/renovate/pull/17193

The current `group:rubyOnRails` preset put the Gems in the `matchPackageNames` array into a single Renovate PR:

```json
{
  "packageRules": [
    {
      "groupName": "Ruby on Rails packages",
      "matchDatasources": [
        "rubygems"
      ],
      "matchPackageNames": [
        "actioncable",
        "actionmailbox",
        "actionmailer",
        "actionpack",
        "actiontext",
        "actionview",
        "activejob",
        "activemodel",
        "activerecord",
        "activestorage",
        "activesupport",
        "railties",
        "rails"
      ]
    }
  ]
}
```

Do you think it's a good idea to add `rack` to the group? I don't use Ruby on Rails or `rack`, so I need your input here. 😉 

> I am wondering if we could remove it entirely, if Renovate is smart enough to detect that we can not bump it due to the major version constraint in `rails`?

I don't know. The quickest way to find out is to try it, and see what happens. 🙈

Comment (renchap) [reply .github/renovate.json5:50]: I dont think so, `rack` is not a gem maintained by Rails (as all the ones you put in your group are), but major Rails versions depend on a specific major `rack` version (through `actionpack`).

The issue here is that we depend on `rack ~> 2.2.7`, as Rails does, but Renovate wants to push an update to `rack > 3`, which is maybe compatible with our Gemfile, but will not be compatible with ActionPack's version (`rack ~> 2.0, >= 2.2.4`)

So I would expect here that Renovate understand that it cant update our `rack` dependency, because one of our dependency's dependency will not support it. 

This is why I chose to have Renovate ignore `rack` major update completely, because we can only bump them to the next major when we upgrade Rails.

File: .github/renovate.json5
Comment (HonkingGoose) [inline .github/renovate.json5:16]: I found a way to remind the maintainers to check the renovate config file before approving any update. I've changed the Dependendy Dashboard issue title, and the header text.

Would this be enough for you to allow major updates via the dashboard? Or do you want me to revert to "ignore major updates completely"?

Comment (renchap) [reply .github/renovate.json5:16]: I did not see those! This will definitely alleviate the issues I had.

But I think it is redundant to have both the header and the title warn to check the config. Could you please only keep the header and add a link to the `renovate.json5` file?

**Review by ClearlyClaire — APPROVED** [review]

**Review by renchap — APPROVED** [review]

**Comment by HonkingGoose** [comment]
@renchap stupid question, but do you run Prettier on incoming PRs? I can't get this project's GitHub Codespaces to work for me (it reverts to the default image, instead of using the Mastodon setup). So I can't check if the styling of my PR follows your "Prettier rules". 🙈

**Comment by renchap** [comment]
> @renchap stupid question, but do you run Prettier on incoming PRs? I can't get this project's GitHub Codespaces to work for me (it reverts to the default image, instead of using the Mastodon setup). So I can't check if the styling of my PR follows your "Prettier rules". 🙈

Yes we do, but it looks like I forgot to add `.json5` to our lint check when added Renovate: https://github.com/mastodon/mastodon/blob/main/package.json#L19

**Comment by HonkingGoose** [comment]
@renchap Oh no... 🙃 I'll let you fix your Prettier config then. 😉

**Comment by shleeable** [comment]
Thanks for this!

**Comment by HonkingGoose** [comment]
Thanks for the reviews and the merge! ❤️

**Pull request merged**
